### PR TITLE
Refactor the page allocator API and expose the frame metadata for page cache

### LINKED ISSUE #1077: Extend the page metadata registry for the page cache
### Summary

This RFC would like to expose the page metadata registry to OSTD users, mainly for the disk cache.

### Context and Problem Statement

The page metadata system, introduced by #836 , has offered an efficient page state registry for page table pages. In this registry, pages, according to it's usage, has been categorized into `Frame` (=`Page<FrameMeta>`, untyped pages), `Page<KernelMeta>` (typed kernel pages) and `Page<PageTablePageMeta>` (typed page table pages). The usages are stored as a field in the common part of the metadata storage to allow dynamic inspection and dispatch of dropping or state maintenance functions.

As refactoring the page cache (FYI, physical memory pages that are presented as a copy of on-disk file contents) becoming the next goal evolving Asterinas, there would be a question how we should maintain the states of page cache pages. The state of a disk cache page is: whether it is not synchronized with the disk content, i.e. dirty. That's it. If the page is not dirty, we can (in certain cases) recycle the page for other usages without writing back the content to the disk. However, recycling disk cache pages would include 3 very tricky parts. The first part is:

1. If we want to recycle it, what about other handles that are currently referencing it? If we want to recycle it, shouldn't we firstly un-map it in all user spaces?

To avoid this problem, the nonsense way is: recycle pages only when the reference count is 0. It is the acutally the solution: aster-nix should do destroy all the referencer to recycle it. This suits our current implementation and our RAII vision really well, since the reference count is the number of the active handles plus the number of virtual memory spaces that maps to it. In order to recycle a disk cache page, aster-nix should 1. un-map all related mappings 2. drop all handles to it.

The second part is:

2. How to tell the dirtiness if it is mapped to the user space? (even worse, shared memory mapped)

With syscall `read` and `write` this is relatively easy. But with memory-mapped files, we must query over the page tables. Luckily, aster-nix could only recycle this page by un-mapping in all the related user spaces. If we collect the dirty bit when doing un-map, we are confident about it's dirtiness after all un-mapping is done.

The third part is:

3. How to enable arbitrary strategy deciding which page to recycle?

This is the most interesting topic. Let us just make up an advanced strategy to see what's required for the state book-keeping. With the idea of LRU (least recently used), we can order the pages in a list. There would be a kernel task that periodically scans over all the address spaces that mapped file-backed pages. When found a accessed/dirty flag, the task clears it and move the page to the end of the list, as "recently used". Therefore, least recently used pages will appear at the front of the list to be recycled.

In order to implement such a strategy, we need to have a global registry of state for pages. The scanning task and the syscall handler would be able to update the dirty state of the pages. Also, the metadata should somehow record the file which the page belongs to. When writing back the content of the page, the task referencing the page should know how to drop other references and which block should we write to.

So then, there is another multiple choice question to be resolve, about how do we code such a registry of states:

 Option 1. Utilize the page metadata system in OSTD by:

    Option 1.1. Allow OSTD users bring their own definition of metadata of frames;

    Option 1.2. Hard code the metadata of frames in OSTD.

 Option 2. Let aster-nix construct a giant global array by itself.

The page metadata system in OSTD was designed to be a centralized, global storage of page states for any page usages. But the way to expose the page metadata system to OSTD users remains unresolved. The major question we want to decide is that: __can we allow OSTD users bring their own definition of frame metadata?__

The answer is, sadly, __NO__. It's because that we are not able to do so. Even if we managed to achieve a certain kind of customizability, such as the `Pod` approach in #1076 or the RTTI approach in #1079 , the restriction will be unacceptable. No exisiting way can enable the complex requirement of disk cache pages presented above. Further details of such restriction will be discussed in the "Alternatives Considered" part of this RFC.

So there would be only 2 choices left. The option 1.2. will decrease OSTD's degree of being general-purpose. The option 2. will be hard to efficiently implement since we don't have unsafe, and wasted about 1/512 of all physical memory since then the metadata system in OSTD will only be useful for a small number of page table pages.

### Proposal

Since I have decided to disallow OSTD users from bringing their own definition of frame metadata, let's inspect the merit and the drawbacks defining a general-purpose frame metadata in OSTD.

Here are the proposal to add frame metadata to the exisiting metadata registry in OSTD, satisfying both the need of disk cache pages and the anonymous pages.

#### The frame metadata

```rust
// in OSTD

/// The metadata of a untyped frames.
pub struct FrameMeta {
  /// If the frame is dirty.
  pub dirty: AtomicBool,
  /// The index to the owning inode.
  /// This field is useless for anonymous pages, but for file backed pages, since the inode can
  /// own a handle to this page, when we want to recycle the page, we should also let the inode
  /// write back the content and drop the page.
  pub inode_index: usize,
  /// We can provide a very efficient unmap method if having this field.
  /// If such efficient methods don't exist, we cannot trivially recycle shared
  /// file backed memory on memory pressure.
  /// It will be really complex in SMP contexts since you need to flush the TLB after unmapping.
  /// So it is a future possibility. We can neglect recycling shared file backed memory.
  pub(crate) mapped_ptnodes: SpinLock<BtreeSet<(RawPageTableNode, PTEIdx, VmSpaceID, Vaddr)>>
}
```

#### Extra APIs to expose

For frame handles, we need an extra API to expose:

```rust
impl Frame {
  /// Get the reference count of the frame, including mappings and handles.
  ///
  /// This method by itself is safe, but using it correctly requires extra care.
  /// Another thread can change the strong count at any time, including potentially
  /// between calling this method and acting on the result.
  pub reference_count(&self) -> u32 {
    self.page.ref_count()
  }
}
```

### Alternatives Considered

In this part, for you who wonders, I give evidences that I tried all I could think to let OSTD users defining their own frame metadata. These attempts have all failed our expectation.

#1076 defined a trait called `FrameMetaExt`, which requires the implementor to be `Pod`. Then we can define metadata in `Pod` in aster-nix. However, `Pod` is an overly-strong requirement that we cannot even use atomic integers, which forbids interior mutability. Without interior mutability the registry is just useless since no one can update the states.

#1079 also introduced a trait `FrameMetaExt` requiring only the dropping callback. It uses RTTI to let OSTD can store objects neglecting the specific definitions of metadata. However, RTTI introduces non-negligible performance overhead and requires heap allocation. Theses are the reasons I ruled out all these implementations.

https://github.com/m-ou-se/rfcs/blob/extern-static/text/0000-externally-definable-statics.md May provide a way to define "existential external types", which would satisfy our need. But it seemed to be too futuristic.

### Additional Information and Resources

[Rust trait object safety](https://doc.rust-lang.org/reference/items/traits.html#object-safety) and [Rust downcast](https://doc.rust-lang.org/stable/std/boxed/struct.Box.html#method.downcast) for the obstacles using RTTI for metadata.

### Open Questions

It is still not decided if the frame metadata's dirty field should be called `dirty`, because that anonymous pages, while would have this information but actually don't need it. It is also undecided that if we should include a drop callback function in the metadata, or let OSTD users maintain a frame to disk device mapping.

### Future Possibilities

Surely we want to implement a "get rid of all mapped address spaces of this frame" with good care.

### PR DESCRIPTION
**Pull request by junyang-zh**
Fix #1077 , but overturning it.

We found out that nevertheless we need to store the reverse mapping from the `Frame` to the `Vmar` and `Vmo`. So the type of the frame metadata just couldn't be decided in OSTD. Here we use polymorphism and [pointer metadata](https://github.com/rust-lang/rust/issues/81513) to tackle with this problem.

Also, the page allocator API and the `Segment` API is refactored to be more advanced. I feel that it's the time.

---

UPDATE: The heap allocation of page allocation is completely eliminated in the API level.
UPDATE: The frame allocation API is restored and the refactor of it is considered out of scope.

---

This PR might go after #1372

### CODE REVIEW

File: ostd/src/mm/frame/options.rs [hunk 1]
```diff
@@ -23,111 +23,62 @@ use crate::{
...
-    ///
-    /// The default value is false.
-    pub fn uninit(&mut self, uninit: bool) -> &mut Self {
-        self.uninit = uninit;
+    /// By default this is `true`. Note that setting this to `false` may
+    /// improve performance, but it definitely introduces security
+    /// vulnerabilities by leaking sensitive information. The caller is
+    /// responsible for writing to all the bits of the allocated frames
+    /// before sharing them with other components.
+    pub fn zero_init(&mut self, zero_init: bool) -> &mut Self {
```
Comment (tatetian) [inline ostd/src/mm/frame/options.rs:48]: This method name is a bit of confusing. How about `zeroed` or `fill_zeros`?

File: ostd/src/mm/frame/options.rs [hunk 2]
```diff
@@ -23,111 +23,62 @@ use crate::{
...
 
-    /// Allocates a contiguous range of page frames according to the given options.
+    /// Allocates a contiguous range of page frames with provided metadata.
     ///
-    /// The returned [`Segment`] contains at least one page frame.
-    pub fn alloc_contiguous<M, F>(&self, mut metadata_fn: F) -> Result<Segment>
+    /// The metadata is initialized with the caller-provided closure
+    /// `metadata_fn`. The closure receives the physical address of the
+    /// page and returns the metadata for the page.
+    pub fn alloc_contiguous<M, F>(&self, nframes: usize, mut metadata_fn: F) -> Result<Segment>
```
Comment (tatetian) [inline ostd/src/mm/frame/options.rs:71]: In the other comment of mine, I suggest renaming `AnyFrame` to `Frame` because we want make the common use cases simple, and advanced use cases possible. In the same spirit, I suggest making the API for allocating frames without user-defined metadata easier. For the simplest use case, the refactored interface is cumbersome to use.

```rust
let single_frame = options.alloc_single(()).unwrap();

let contiguous_segment = options.alloc_contiguous(10, |_| ()).unwrap();
```

Let's say that you have adopted the name `Frame` (currently known as `AnyFrame`), `FrameEx<M>` (currently known as `Frame`). And similarly, `Segment` and `SegmentEx`. Then, the allocation API can be revised to the following form.

```rust
impl FrameAllocOptions {
    pub fn alloc_single(&self) -> Result<Frame> { ... }

    pub fn alloc_single_ex<M>(&self) -> Result<FrameEx<M>> { ... }

    pub fn alloc_contiguous(&self, nframes: usize) -> Result<Segment> { ... }

    pub fn alloc_contiguous_ex<M, F>(&self, nframes: usize, mut metadata_fn: F) -> Result<SegmentEx<M>> { ... }
}
```

File: ostd/src/mm/page/cont_pages.rs [hunk 3]
```diff
@@ -100,3 +171,21 @@ impl<M: PageMeta> From<ContPages<M>> for Vec<Page<M>> {
...
 }
+
+impl<M: PageMeta> Iterator for ContPages<M> {
+    type Item = Page<M>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.range.start < self.range.end {
+            // SAFETY: each page in the range would be a handle forgotten
+            // when creating the `ContPages` object.
+            let page = unsafe { Page::<M>::from_raw(self.range.start) };
```
[the remark this answers is not in the document]
Comment (junyang-zh) [reply ostd/src/mm/page/cont_pages.rs:182]: This is correct since the semantic of cont_pages.next(), though weird, consumes the first page and hands the ownership of the first page to the caller.

File: kernel/comps/block/src/impl_block_device.rs [hunk 4]
```diff
@@ -72,14 +76,18 @@ impl dyn BlockDevice {
     }
 
     /// Synchronously writes one block indicated by the `bid`.
-    pub fn write_block(&self, bid: Bid, frame: &Frame) -> Result<BioStatus, BioEnqueueError> {
+    pub fn write_block(&self, bid: Bid, frame: &AnyFrame) -> Result<BioStatus, BioEnqueueError> {
```
[the remark this answers is not in the document]
Comment (junyang-zh) [reply kernel/comps/block/src/impl_block_device.rs:79]: In this PR `Frame` (which is `Frame<()>`) appeared 63 times, `AnyFrame`appeared 118 times `CachePage` (which is `Frame<CachePageMeta>`) appeared 12 times.

From my perspective, `AnyFrame` is neither the most encouraged type now nor the most prevalent type in the future:

 - A frame should have it's purpose (anonymous/page cache/buffer) and their accessing behavior will be distinct. Almost all frames will have metadata (anonymous pages will gain it's reverse mapping in the future). So when the metadata of a frame is accessed, `AnyFrame` is not the best option. Metadata will be frequently accessed. For this reason, any internal APIs that are dedicated to one usage (such as VMO, which holds only page cache pages), will eventually turn to typed frame.
 - If `Frame` represents `AnyFrame` rather than `Frame<()>`, the allocation scheme, as discussed in https://github.com/asterinas/asterinas/pull/1287#discussion_r1774375098, makes the API far harder to read and semantically indiscreet. The metadata of an allocated `Frame` can be `()`, but the user cannot tell it from the API. Also, the `Ex` suffix makes less intuitive names. The user cannot tell directly that the extension means attaching metadata other than `()`. The naming dilemma just indicates a bad design.

Also `AnyFrame` is a good name indicating that the handle points to anything that is a frame. The methods of `AnyFrame` can be applied to "any frame".

To sum up, I support leaving the naming as is in this PR.

**Comment by junyang-zh** [comment]
prefer #1724
